### PR TITLE
Update docs of `set_if_neq` and `replace_if_neq`

### PR DESCRIPTION
**Pull request by RobWalt**
# Objective

- ~~This PR adds more flexible versions of `set_if_neq` and `replace_if_neq` to only compare and update certain fields of a components which is not just a newtype~~
- https://github.com/bevyengine/bevy/pull/12919#issuecomment-2048049786 gave a good solution to the original problem, so let's update the docs so that this is easier to find

## Solution

- ~~Add `set_if_neq_with` and `replace_if_neq_with` which take an accessor closure to access the relevant field~~

---

In a recent project, a scenario emerged that required careful consideration regarding change detection without compromising performance. The context involves a component that maintains a collection of `Vec<Vec2>` representing a horizontal surface, alongside a height field. When the height is updated, there are a few approaches to consider:

1. Clone the collection of points to utilize the existing `set_if_neq` method.
2. Inline and adjust the `set_if_neq` code specifically for this scenario.
3. (Consider splitting the component into more granular components.)

It's worth noting that the third option might be the most suitable in most cases.

A similar situation arises with the Bevy internal Transform component, which includes fields for translation, rotation, and scale. These fields are relatively small (`Vec3` or `Quat` with 3 or 4 `f32` values), but the creation of a single pointer (`usize`) might be more efficient than copying the data of the other fields. This is speculative, and insights from others could be valuable.

Questions remain:

- Is it feasible to develop a more flexible API, and what might that entail?
- Is there general interest in this change?

There's no hard feelings if this idea or the PR is ultimately rejected. I just wanted to put this idea out there and hope that this might be beneficial to others and that feedback could be valuable before abandoning the idea.

### CODE REVIEW

**Comment by SkiFire13** [comment]
> When the height is updated, there are a few approaches to consider:

Note that there's another alternative which is also very similar to the approach proposed in the PR. You can use `map_unchanged` to perform the mapping step and then call `set_if_neq` on the mapped `Mut`.

In fact I'm pretty sure (hasn't actually tried though) that you can implement `set_if_neq_with` as `self.map_unchanged(accessor_f).set_if_neq(value)`

**Comment by RobWalt** [comment]
Thanks for the hint, I didn't know that! Then maybe that should be the way to go and there's no real need for `set_if_neq_with`. It might make sense to update the docs to point other people to this solution here, though. I'll do that once I find the time

**Comment by alice-i-cecile** [comment]
@SkiFire13 can I get your review here?

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/change_detection.rs [hunk 1]
```diff
@@ -131,6 +131,11 @@ pub trait DetectChangesMut: DetectChanges {
     /// This is useful to ensure change detection is only triggered when the underlying value
     /// changes, instead of every time it is mutably accessed.
     ///
+    /// If you're dealing with non-trivial structs which have multiple fields of non-trivial size,
+    /// then consider applying a [`map_unchanged`](Mut::map_unchanged) beforehand to prevent the
+    /// need for unnecessary copying and cloning. See the docs of
+    /// [`map_unchanged`](Mut::map_unchanged) for an example
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/change_detection.rs:137]: I would specify that the idea is that after `map_unchanged` only the relevant field needs to be set, not the whole value. Also, concluded the last sentence with a dot.

```suggestion
    /// If you're dealing with non-trivial structs which have multiple fields of non-trivial size,
    /// then consider applying a [`map_unchanged`](Mut::map_unchanged) beforehand to allow
    /// changing only the relevant field and prevent unnecessary copying and cloning.
    /// See the docs of [`map_unchanged`](Mut::map_unchanged) for an example.
```

File: crates/bevy_ecs/src/change_detection.rs [hunk 2]
```diff
@@ -181,6 +186,11 @@ pub trait DetectChangesMut: DetectChanges {
...
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/change_detection.rs:192]: I would do the same here.

File: crates/bevy_ecs/src/change_detection.rs [hunk 3]
```diff
@@ -131,6 +131,11 @@ pub trait DetectChangesMut: DetectChanges {
     /// This is useful to ensure change detection is only triggered when the underlying value
     /// changes, instead of every time it is mutably accessed.
     ///
+    /// If you're dealing with non-trivial structs which have multiple fields of non-trivial size,
+    /// then consider applying a [`map_unchanged`](Mut::map_unchanged) beforehand to prevent the
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/change_detection.rs:135]: Linking to only `Mut::map_unchanged` is not totally correct, as this situation can happen with `MutUntyped`, `ResMut` and `NonSendMut` too. However I don't see a way to fix this without verbosely listing all of them, so ideas are welcome. If nobody finds a better way to express this I would keep it like this.

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/change_detection.rs:135]: I think this is adequately clear: users who are at this level of sophistication will be able to piece this together.

Comment (RobWalt) [reply crates/bevy_ecs/src/change_detection.rs]: I hope I handled this adequately to make everyone here happy

**Review by SkiFire13 — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Thank you both: this is a helpful note.

**Pull request merged**